docs/blueprints/deletion-performance.md

```diff
@@ -0,0 +1,188 @@
+- It doesn't add more load to the satellite, but we cannot trust in the uplink in deleting the pieces or informing the non-deleted ones.
+
+The alternative approaches (2) and (3) are similar.
+
+Approach (2) has the advantage of guaranteeing less garbage left on the storage nodes at the expense of requiring more network traffic.
+
+Approach (3) requires less network traffic, but it may not require less computation considering that we may need to encrypt the data sent through UDP.
+
+Approach (3) may get rid of the satellite garbage faster.
+
+Both approaches, (2) and (3), present the problem of increasing garbage on the satellite side.
+
+Taking one of these approaches will require a study on how to keep the less amount of garbage as possible.
+
+## Implementation
+
+### (1) Storage nodes:
+
+1. Adapt protocol buffers definitions for delete operation.
+       Satellite doesn't have to send order limits; it only has to send piece ID.
+1. Adapt the delete endpoint to receive requests from the satellite.
+
+### (2) Satellite:
+
+1. Implement delete request logic with backpressure mechanism to only confirm the operation when certain amount of storage nodes confirm successful.
+   It's like what we do for upload long-tail cancellations.
+1. Adapt protocol buffers definitions for delete operation.
+   The current uplink RPC requests are:
+
+        1. `BeginDeleteObject` - Retreives the _stream ID_.
+        1. `ListSegments` - Uses _stream ID_  for retrieving the position index of all the segments.
+        1. `BeginDeleteSegments` - Uses the _stream ID_ and position index for retrieving a list of _addressed order limits_.
+
+   Because uplink won't send the delete requests to the storage nodes, the delete operation we can simplify it with one satellite request. The satellite will respond, with an empty body, when the deletion ends.
+
+### (3) Uplink:
+
+1. Change logic to not send delete requests to storage nodes.
+1. Uplink `rm` command will wait until satellite responds.
+
+### Considerations
+
+If we plan to release the feature in several steps:
+
+1. Implement and independently release (1) and (2) without removing the logic of the current functionality.
+1. Implement and release (3).
+1. Announce when previous versions will stop to work properly.
+1. Remove and independently release the delete old logic from the storage node and satellite.
+
+
+## Open issues (if applicable)
+
+1. Should we track how much each storage node is storing extra due not sending deletes? For the storage nodes that accumulate too much garbage, we could send garbage collection request outside of the normal schedule.
+1. Discuss backward-incompatibility that we may introduce when adapting the protocol buffer definitions in the storage node and satellite side. Should we reuse the current definition as much as possible although it isn't ideal?
+1. How many storage nodes must confirm the deletion successful to allow the satellite to return a response to the uplink?
```

Comment:
Why can't we just respond to the uplink as soon as we delete from metainfo and wait for the storagenode deletion responses independently? That way from the uplink perspective the delete is instantaneous and we do the deletion from the storagenodes behind the scenes?

Reply:
@jtolds was not very keen about doing so in the first attempt. If after doing these architectural change we find out that we are still too slow deleting objects, then we'll think in a solution which can involve the one that you mentioned.

If I don't remember bad, he commented that backpressure is something that always happens in a distributed system and we have to deal with it, however, don't take my word.